### PR TITLE
Add IMessage to populate additional headers in RawMessage.

### PR DESCRIPTION
**Pull request by Scot-Forbes-Exp**
- Adds a new IMessage<T> interface, where T is the type of the message (as if you had passed it directly to Send/Publish.
- Adds a RawMessage Create overload that copies headers from the IMessage to the RawMessage
- Adds Send/Publish overloads that take IMessage<T> instead of object
- Adds a transport test to verify the preservation of custom headers in the primary queue.
- Transport test passes for RabbitMQ and DefaultTransport

**What is not included/verified:**
- Transport test was not verified on MSMQ since I do not have it installed locally. My assumption is that it behaves similarly enough to the DefaultTransport in its handlers that it should be fine :)
- There is currently no way to read and respond to the additional headers at the transport level, for example to set the priority on the message. That will be a future pull request. The headers *can* currently be used in the message filters, however, since it is part of the message envelope.
- No new unit tests. Couldn't think of any to write. Suggestions are welcome.

**Some open questions:**
- How should we handle the edge case of the IMessage specifying a "built-in" header like "Id" or "ReplyTopic"? The current behavior is to silently ignore them.

### CODE REVIEW

**Comment by jgorman-exp** [comment]
This works for me.  Using the existing Headers property is an interesting option to me.  The "headers" paradigm seems to make sense for a message envelope, and I could imagine there being confusion around what to put in Headers vs Metadata.  Ok with keeping Metadata, though.

**Comment by Joholland** [comment]
👍 This is your repo now, you don't need me :D

File: tests/Wave.Core.Tests/TransportTestBase.cs [hunk 1]
```diff
@@ -226,6 +258,14 @@ protected void RunBlocking(Action<ManualResetEvent> action, TimeSpan timeout)
         }
 
         [Serializable]
-        protected class TestMessage { }       
+        protected class TestMessage { }
+        
+        [Serializable]
+        protected class TestMessageWithMetadata : IMessage<TestMessage>
```
Comment (Scot-Forbes-Exp) [inline tests/Wave.Core.Tests/TransportTestBase.cs]: Decided I didn't need this class but forgot to remove it.

File: src/Wave.Core/RawMessage.cs [hunk 2]
```diff
@@ -119,6 +120,28 @@ public static RawMessage Create(object message)
...
+            var rawMessage = Create(message.Content);
+
+            var existingHeaders = rawMessage.Headers;
+            var additionalHeaders = message.Headers
+                ?? Enumerable.Empty<KeyValuePair<string, string>>();
+
+            foreach (var header in additionalHeaders)
+            {
+                // TODO: Throw if trying to override built-in headers instead of silently ignoring?
+                //       Or do we allow overwriting?
```
Comment (jgorman-exp) [inline src/Wave.Core/RawMessage.cs]: Current behavior makes sense to me...  Ignore any sets on reserved headers.
